react/the_react_ecosystem/react_router.md

```diff
@@ -0,0 +1,589 @@
+      <hr />
+      <button type="button" onClick={login}>
+        Login
+      </button> <button type="button" onClick={logout}>
+        Logout
+      </button>
+      <hr />
+      <p>Here are some examples of links to other pages</p>
+      <nav>
+        <ul>
+          <li>
+            <Link to="protected">Protected page</Link>
+          </li>
+        </ul>
+      </nav>
+    </div>
+  );
+};
+
+export default App;
+~~~
+
+Now, let us create a component that should only be allowed to be seen by logged-in users. Let us call it `Protected.js`
+
+~~~jsx
+import { Link } from "react-router-dom";
+
+const Protected = () => {
+  return (
+    <div>
+      <p>Hey! You should only see this if you are logged in!</p>
+      <p>
+        <Link to="/">Click here</Link> to go back to the home page
+      </p>
+    </div>
+  );
+};
+
+export default Protected;
+~~~
+
+Now, We can pass the routes conditionally to the `createBrowserRouter` methods based on if the user is logged in. Else, we can use another one of React Router's components, `<Navigate>` to redirect the user back to another page, in this instance, let it be Popeye's profile page.
+
+~~~jsx
+import { useState } from "react";
+import {
+  createBrowserRouter,
+  Navigate,
+  RouterProvider,
+} from "react-router-dom";
+import App from "./App";
+import Profile from "./Profile";
+import ErrorPage from "./ErrorPage";
+import Protected from "./Protected";
+
+const Router = () => {
+  const [isLoggedIn, setIsLoggedIn] = useState(false);
+
+  const routes = (condition) => [
+    {
+      path: "/",
+      element: <App isLoggedIn={isLoggedIn} setIsLoggedIn={setIsLoggedIn} />,
+      errorElement: <ErrorPage />,
+    },
+    {
+      path: "profile/:name",
+      element: <Profile />,
+    },
+    {
+      path: "protected",
+      element: condition ? <Protected /> : <Navigate to="/profile/popeye" />,
+    },
+  ];
+
+  const router = createBrowserRouter(routes(isLoggedIn));
+
+  return <RouterProvider router={router} />;
+};
+
+export default Router;
+~~~
+
+Here, the routes is a function that accepts a condition to see if the user is logged in. It uses that to generate a configuration of routes which are then passed on to the `createBrowserRouter` method to finally make the router.
+
+### Assignment
+
+<div class="lesson-content__panel" markdown="1">
+
+1.  [This article](https://bholmes.dev/blog/spas-clientside-routing/) by Ben Holmes goes through a lot of the routing concepts concisely.
+2.  Go and add a few new routes to the application we created above; playing around with it is the best practice. Consider deleting it completely and rewriting it for practice.
+3.  The [React Router tutorial](https://reactrouter.com/en/6.7.0/start/tutorial) goes through a lot of the stuff discussed in this lesson and much more. Have a read into it, and ask about, if you don't understand something. You are not expected to understand most of it in your first read.
```

Comment:
This tutorial goes over a lot of extra stuff not covered in the lesson and I'm not sure if learners need to go through all of that information right now. We should narrow down what topics the learners need to cover in that tutorial.

Reply:
What are the specifics that should be pointed out? I think we can assume that loaders, actions and mutations, and a few more components can be pretty excessive. But I'm not sure how to tell the learner to skip these.
Guess we can explicitly mention the sections necessary and leave the learner to read the rest if they want? But there are quite a few sections, so I don't know how the can be added in a concise manner other than just listing them out.

Reply:
Yeah, we just ask them to read from "Adding A Router" to "Nested Routes" section


![image](https://user-images.githubusercontent.com/85733202/215097822-494c5cd4-44ec-43c1-bfe0-8e9502451858.png)